analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Invocation.cs

```diff
@@ -33,10 +35,36 @@ internal static class Invocation
                 : context.State;
 
         public static ProgramState Process(SymbolicContext context, IArgumentOperationWrapper argument) =>
-            argument.Parameter is not null      // __arglist is not assigned to a parameter
-            && argument.Parameter.RefKind != Microsoft.CodeAnalysis.RefKind.None
-            && argument.Value.TrackedSymbol() is { } symbol
-                ? context.State.SetSymbolValue(symbol, null)
-                : context.State;
+            ProcessArgument(context.State, argument) ?? context.State;
+
+        private static ProgramState ProcessArgument(ProgramState state, IArgumentOperationWrapper argument) =>
+            argument switch
+            {
+                { Parameter: null } => null, // __arglist is not assigned to a parameter
+                { Parameter.RefKind: not RefKind.None, Value: { } value } when value.TrackedSymbol() is { } symbol =>
+                    // The argument is passed by some kind of reference, so we need to forget all we knew about it.
+                    state.SetSymbolValue(symbol, null),
+                { Parameter: { } parameter } when parameter.GetAttributes() is { Length: > 0 } attributes =>
+                    // Learn from parameter nullable annotations
+                    ProcessArgumentAttributes(state, argument, attributes),
+                _ => null,
+            };
+
+        private static ProgramState ProcessArgumentAttributes(ProgramState state, IArgumentOperationWrapper argument, ImmutableArray<AttributeData> attributes)
+        {
```

Comment:
Let's use `FirstOrDefault(IsValidatedNotNullAttribute) is { } attribute`. That will remove bunch of lines and enable arrow method with ternary inside.

Reply:
As a user, I expect all of these guards to work, and it is likely to have these for conciseness.
```cs
public void NotNull<T>([ValidatedNotNull] T arg);
public void NotNull<T1, T2>([ValidatedNotNull] T1 arg1, [ValidatedNotNull] T2 arg2);
public void NotNull<T1, T2>([ValidatedNotNull] T1 arg1, [ValidatedNotNull] T2 arg2, [ValidatedNotNull] params object[] others);

NotNull(o1, o2, o3 o4, o5);
```